retrieval/discovery/kubernetes: fix cache state unknown behavior

brancz opened the pull request:
The `OnDelete` handler can under certain circumstances [return a `DeletedFinalStateUnknown`](https://github.com/kubernetes/client-go/blob/843f7c4f28b1f647f664f883697107d5c02c5acc/1.5/tools/cache/controller.go#L146-L149) object. In that case we have a stale cache, but will only return to a correct cache once the refresh period has passed, so we keep our cache and use it on a best effort basis until then, without `panic`ing here.

@fabxc @brian-brazil 

Fixes #2178

jimmidyson commented on retrieval/discovery/kubernetes/pod.go:
```diff
@@ -74,7 +74,10 @@ func (p *Pod) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 		},
 		DeleteFunc: func(o interface{}) {
-			send(&config.TargetGroup{Source: podSource(o.(*apiv1.Pod))})
+			pod, ok := o.(*apiv1.Pod)
```
We only need the namespace & name which is available in the (albeit stale) `DeletedFinalStateUnknown.Obj` field?

brancz replied:
Yep, [sounds like](https://github.com/kubernetes/client-go/blob/843f7c4f28b1f647f664f883697107d5c02c5acc/1.5/tools/cache/delta_fifo.go#L631-L634) the type cast should be safe on the inner Obj. Will update.

jimmidyson commented on retrieval/discovery/kubernetes/endpoints.go:
```diff
@@ -89,6 +89,11 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+				o = s.Obj
+			}
+
 			send(&config.TargetGroup{Source: endpointsSource(o.(*apiv1.Endpoints).ObjectMeta)})
```
Is it worth adding a type assertion test to make sure there's no chance of a panic here?

fabxc replied:
Yea, I think we should stick with proper validation everywhere to avoid future panics. The underlying code is not easy enough to reason about to rely on contract here as the issue has shown.

https://github.com/kubernetes/kubernetes/pull/34694/files

fabxc replied:
I generally agree that panicking on contract violation is fine. But this is a third party dependency that's too large and too quickly changing that we can realistically validate changes and correctness on every vendoring update.

And we cannot really risk crashing our users servers just because a dependency of the SD integration is not working as we expect.

brancz replied:
Updated.

fabxc commented:
Can we embed the check in the `send()` functions?
Shouldn't work without modifications for the non-delete case as well and make those a bit safer too.

brancz commented:
sgtm, also makes it a bit more readable

brancz commented:
Actually we can't, as the send function expects a `TargetGroup`, and as the `DeleteFunc` needs to build an empty `TargetGroup`, we can't build the `TargetGroup` in the send functions. What I'll do is have a cast/convert function that does this before building the actual `TargetGroup`.

brancz commented:
Extracted the conversion/casting and added some test to cover the reported bug in terms of possible regressions. Re-review would be appreciated @fabxc @jimmidyson

fabxc commented:
👍

brancz commented on retrieval/discovery/kubernetes/endpoints.go:
```diff
@@ -83,23 +83,50 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 		obj, exists, err := e.endpointsStore.Get(ep)
 		if exists && err != nil {
-			send(e.buildEndpoints(obj.(*apiv1.Endpoints)))
+			eps, err := convertToEndpoints(obj)
```
Removed this as discussed offline with @fabxc , because we are relying in many places that at least the store stores consistently typed objects.
